crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::path::Path;
+
+use failure::{bail, format_err};
+use log::info;
+use wasi_common::preopen_dir;
+use wasmtime_wasi::old::snapshot_0::Wasi as WasiUnstable;
+use wasmtime_wasi::{Wasi, WasiCtxBuilder};
+
+/// WasiRuntime provides a WASI compatible runtime. A runtime should be used for
+/// each "instance" of a process and can be passed to a thread pool for running
+pub struct WasiRuntime {
+    /// binary module data to be run as a wasm module
+    module_data: Vec<u8>,
+    /// key/value environment variables made available to the wasm process
+    env: HashMap<String, String>,
+    /// the arguments passed as the command-line arguments list
+    args: Vec<String>,
+    /// a hash map of local file system paths to optional path names in the runtime
+    /// (e.g. /tmp/foo/myfile -> /app/config). If the optional value is not given,
+    /// the same path will be allowed in the runtime
+    dirs: HashMap<String, Option<String>>,
+}
+
+impl WasiRuntime {
+    /// Creates a new WasiRuntime
+    ///
+    /// # Arguments
+    ///
+    /// * `module_path` - the path to the WebAssembly binary
+    /// * `env` - a collection of key/value pairs containing the environment variables
+    /// * `args` - the arguments passed as the command-line arguments list
+    /// * `dirs` - a map of local file system paths to optional path names in the runtime
+    ///     (e.g. /tmp/foo/myfile -> /app/config). If the optional value is not given,
+    ///     the same path will be allowed in the runtime
+    /// * `log_file_location` - location for storing logs
+    pub fn new<M: AsRef<Path>>(
+        module_path: M,
+        env: HashMap<String, String>,
+        args: Vec<String>,
+        dirs: HashMap<String, Option<String>>,
+    ) -> Result<Self, failure::Error> {
+        let module_data = wat::parse_file(module_path)?;
+
+        // We need to use named temp file because we need multiple file handles
```

Comment:
I think I forgot to move this comment to where we are constructing the temp files in the Provider as it no longer makes sense here